feat: JavaScript related macros and more (#240)
FyraLabs/devdocs: pages/terra/SRPM.mdx

### Node.js Macros
Macros for using [Node.js](https://nodejs.org). These require use of the [Fedora Node.js macros](https://docs.fedoraproject.org/en-US/packaging-guidelines/Node.js/#_macros) in the `nodejs-packaging` package.

You MUST either set `%global npm_name pkgname` or use `%npm_prep -n pkgname` when packaging Node modules directly. This MUST be defined as the canonical name of the module as listed on the NPM Registry.

#### %_nvm_dir
The location in the `%{rpmbuilddir}` what will be passed to `$NVM_DIR`.

Expands to: `%{rpmbuilddir}/.nvm`

#### %__nvm
Calls NVM (Node Version Manager) while setting `$NVM_DIR`. Only used if vendoring Node.js.

Expands to: `/usr/bin/env NVM_DIR=%{_nvm_dir} /usr/bin/nvm`

#### %vendor_nodejs
SHOULD only be used if Node.js packages available from the repos are not usable for some reason.

You MUST add `nvm` as a build requirement to the spec if using this macro.

Available flags:
- `-v` is REQUIRED for this macro to pass the version to NVM. Accepts a full version, the major for the latest of a major version of Node.js (e.g., `24`), or `node` to use the latest version of Node.js available.

> Usecase examples:

Reviewer: Could you add periods in this list?

- Fedora has not yet packaged the correct version of Node.js or the required version is older than what is supplied
  - This means that packaged Node modules will probably not meet their runtime requirements, but Electron apps may be fine
- EL does not yet have the correct version of Node.js
  - This poses the same concern as above

For this reason whatever you are packaging SHOULD be tested locally upon a successful build to see if it meets any runtime issues.

> What `%vendor_nodejs` does:

Reviewer: Periods here too

- `-b` must be used when the app's executable differs from the name of the package.
- `-d` can be used to install the app under a different directory than the name of the package.
- `-i` must be used when the app's desktop file points to an icon name different from the name of the package.

> - `-l` can be used to fetch all licenses including the bundled ones and put them in a directory called `bundled_licenses`. You can then put them in the license directory using `%license bundled_licenses/*`

Reviewer: ```suggestion
- `-l` can be used to fetch all licenses including the bundled ones and put them in a directory called `bundled_licenses`. You can then put them in the license directory using `%license bundled_licenses/*` in the %files section.
```

Which simply means Electron supports x86_64, x86, aarch64/ARM64, and armv7l.

## Go
These macros assist with building Go projects.

### %go_task
Should be used in place of `go-task`.

## JavaScript/JavaScript Package Managers
These are macros to assist with the general packaging of applications that use JavaScript runtimes, such as Node.js or Bun.
[TODO]: make this sound not stupid
[TODO]: descriptions for the sub-sections
[TODO]: examples/links to examples of how to use the macros here (at least the common ones)

### Node.js Macros

> Macros for Node.js itself. These require use of the existing Node.js macros in Fedora's `nodejs-packaging` package.

Reviewer: Link to upstream guidelines for these macros?

#### %_nvm_dir
Sets the NVM dir to a location in the `%{rpmbuilddir}`.

Expands to: `%{rpmbuilddir}/.nvm`

#### %__nvm
Calls NVM (Node Version Manager) while setting `$NVM_DIR`. Only used if vendoring Node.js.

Expands to: `/usr/bin/env NVM_DIR=%{_nvm_dir} /usr/bin/nvm`

### Node Module Macros
"Node modules" refers to JavaScript packages run using a JavaScript runtime (usually Node.js), almost always hosted on the [NPM Registry](https://www.npmjs.com/).

#### %npm_prep
Prepare a Node module for build steps.

Available flags:
- `-n`: Use this if you have not defined `%{npm_name}` as a global macro.

#### %fetch_node_tests
Fetch self-tests from a Git repo. MUST include all necessary test files and folders. If used, SHOULD be used in `%prep`.

Available flags:
- `-u` Define the URL if the URL for the package is not detectable as a clonable URL.

All necessary test files and folders SHOULD be listed after the macro. Test folders MUST start with `/`.

Example usage:
```rpmspec
%prep
%npm_prep
%fetch_node_tests -u https://github.com/dev/module.git /tests .eslintrc
```

#### %npm_install
Install a Node module to `%{npm_sitelib}`.

> #### %npm_test

Reviewer: ```suggestion
Accepts a `-o` flag as a shorthand for `--out`.
```

Author: Did you mean to remove the header with this? LOL

Which simply means Electron supports x86_64, x86, aarch64/ARM64, and armv7l.

## Go
These macros assist with building Go projects.

> #### %goprep_online

Reviewer: This should use three #'s, not 4. And look like you accidentally removed %go_task?

(Line 177 if GitHub mobile screws this up)

Author: I moved it to general macros because it isn't actually relevant to Go, basically same reason for moving it to `macros.anda`.

Reviewer: Ohh I see whoops

NPMFLAGS="${NPMFLAGS:-%{npm_buildflags} }" ; export NPMFLAGS
```

This ensures the environment variables are explicitly exported and are more likely to be picked up by all shell processes. It also creates an environment variable called `$NPMFLAGS` which can be passed to NPM in lieu of the macro `%{npm_buildflags}` to force our config, if necessary.

This should only be needed in the event you are using [`%__npx`](#__npx), [`%__pnpx`](#__pnpx), [`%__bunx`](#__bunx), or [`%__yarn_dlx`](#__yarn_dlx).

## Electron
These macros assist with the general packaging of applications that use [Electron](https://www.electronjs.org).

#### %electronmeta
Set up all basic requirements including build macros for building an Electron app.

Does <u>not</u> install the JavaScript package managers used for builds (e.g., `nodejs-npm`, `bun`, `pnpm`, `yarnpkg`, `yarnpkg-berry`). These must still be specified in the `BuildRequires:` manually.

Available flags:
- `-a` will pull in additional build dependencies often needed particularly when building with NPM.
  - Currently, this includes `nodejs-devel`, `nodejs-packaging`, and `typescript`.

> Rundown of what all `%electronmeta` does:

Reviewer: Some of these bullets have periods, some dont. Can we unify that?

Author: Sure. With or without periods?

Reviewer: Hmm. Check whatever my PR has, I think I have them.

Also, make sure all section info lines have periods (the sentence descriptions under the macro sections). I saw some missing there too.

1. The JavaScript package manager cache directories are placed in [`%{rpmbuilddir}`](#rpmbuilddir) because `$HOME` can be different depending on if a build is in Mock or RPM Build.
  - This is very important for consistent macro behavior as well as if the cache must be modified or deleted during build for any reason.
2. All macros calling the package managers also call NPM environment variables because most of them are using build dependencies from the NPM Registry, and some even wrap NPM commands with their own. This ensures consistent NPM behavior.

### NPM Macros
Macros for using [NPM](https://www.npmjs.com).

#### %_npm_cache_dir
Expands to: `%{rpmbuilddir}/.npm`

#### %npm_common_envvars
**NOTE:** NPM environment variables written in uppercase will be __ignored__ if the project sets different ones explicitly. This is intentional behavior which these macros do not override in order to avoid build issues. For more information, please see [this issue](https://github.com/npm/npm/issues/14528) in the original NPM GitHub repository.
  - If for any reason you need to hardcode an environment override, use a flag passed to NPM or manually set the environment variable in all lowercase.

>   - For more information on NPM config flags and environment variables, please see: https://docs.npmjs.com/cli/v11/using-npm/config

Reviewer: ```suggestion
  - For more information on NPM config flags and environment variables, please go [here](https://docs.npmjs.com/cli/v11/using-npm/config).
```

Author: Funny thing, I did have a period here originally, but because this ends in a link it seemingly caused issues or the period would just...not show up?

Reviewer: Hmmm ok, lemme change the edit request.

Reviewer: Ok there

Expands to: `NPM_CONFIG_USERCONFIG=%{rpmbuilddir}/.npmrc NPM_CONFIG_GLOBALCONFIG=%{rpmbuilddir}/npmrc NPM_CONFIG_CACHE=%{_npm_cache_dir} NPM_CONFIG_LOGLEVEL=error NPM_CONFIG_FUND=false NPM_CONFIG_UPDATE_NOTIFIER=false NO_UPDATE_NOTIFIER=1 NPM_CONFIG_COLOR=always NPM_CONFIG_INIT_MODULE=%{rpmbuilddir}/.npm-init.js`

What these do:

> - Forces all configuration files into the `%{rpmbuilddir}` so they are easily locatable and to prevent attempting to write to the host file system for global config

Reviewer: ```suggestion
- Forces all configuration files into the `%{rpmbuilddir}` so they are easily locatable and to prevent attempting to write to the host file system for global config.
```

Available flags:
- `-a` can be used to trust all dependencies, otherwise simply list the dependencies you want to trust after the command.

Expands to: `%{__bun} pm trust`

### PNPM Macros
Macros for using [PNPM](https://pnpm.io).

#### %_pnpm_home
Where PNPM files and cache are stored.

Expands to: `%{rpmbuilddir}/.pnpm`

#### %_pnpm_store
Where Node modules installed by PNPM usually go. This will usually not be produced when using PNPM as a build tool, but exists for rare builds it may be created in.

Expands to: `%{_pnpm_home}/store`

#### %pnpm_common_envvars
Sets the NPM evironment variables and the PNPM home dir.

Expands to: `%{npm_common_envvars} PNPM_HOME=%{_pnpm_home}`

#### %__pnpm

Expands to: `/usr/bin/env %{pnpm_common_envvars} %{?with_vendored_pnpm:%{rpmbuilddir}/pnpm/pnpm}%{!?with_vendored_pnpm:/usr/bin/pnpm}`

#### %__pnpx
Remotely execute Node modules using PNPX.

Expands to: `/usr/bin/env %{pnpm_common_envvars} /usr/bin/pnpx`

#### %vendor_pnpm
SHOULD only be used if PNPM is not usable/available from the available repositories.

Available flags:
- `-v` MAY be used to define a desired version of PNPM.

> Usecase examples:

Reviewer: Periods

- Fedora's currently packaged PNPM is too old for the build
- You want to make the package build on EL (does not have PNPM)
  - In this case, you SHOULD wrap `%vendor_pnpm` in `%if` macros such as:
  - ```rpmspec
     %if %{defined rhel}
     %vendor_pnpm
     %endif
    ```

#### %pnpm_audit

> Audit with PNPM if package uses bundled Node dependencies. For use in the `%check` section

Reviewer: ```suggestion
Audit with PNPM if package uses bundled Node dependencies. For use in the `%check` section.
```

Expands to: `/usr/bin/license-checker --limitAttributes licenses`.
Above macro, but limits the output to only the license.

Accepts an `-o` flag as a shorthand for `--out`.

Example output:

```r
/usr/bin/license-checker --limitAttributes licenses
└─ @fyralabs/devdocs@@0.0.1
   └─ licenses: MIT
```

SHOULD be used if packaging Node modules.

Example usage:

```
%{npm_license} --out LICENSE.modules
```

> OR:

Reviewer: Link to this macros documentation?

Author: Which line was this for LOL

Reviewer: Oh uhh was a cmake thing I think, I'll find it soon.

Reviewer: I forgot maybe it was for line 222 but you can resolve this whoops.

Author: Do you think it was maybe the one referencing `%set_build_flags` from Fedora's Red Hat macros??

Reviewer: Ohh yeah possibly. Ideally we link to macros whenever they're like, one-off referenced (link to our RPM section about then maybe?)

- Disables update notifications since we are using a packaged version of NPM unless using `%vendor_nodejs`
- Sets color to "always" for more easily readable logs

#### %npm_buildflags
Explicit NPM configuration flags of the environment variables set by the Macros.

Expands to: `--userconfig=%{rpmbuilddir}/.npmrc --globalconfig=%{rpmbuilddir}/npmrc --cache=%{rpmbuilddir}/.npm --loglevel=error --fund=false --update-notifier=false --color=always --init-module=%{rpmbuilddir}/.npm-init.js`

#### %set_node_build_flags
Set the flags and environment variables used for JavaScript package manager builds.

Based on Fedora's `%set_build_flags`.

Expands to:
```rpmspec
NPM_CONFIG_USERCONFIG="%{rpmbuilddir}/.npmrc" ; export NPM_CONFIG_USERCONFIG ; \
NPM_CONFIG_GLOBALCONFIG="%{rpmbuilddir}/npmrc" ; export NPM_CONFIG_GLOBALCONFIG ; \
NPM_CONFIG_CACHE="%{_npm_cache_dir}" ; export NPM_CONFIG_CACHE ; \
NPM_CONFIG_LOGLEVEL="error" ; export NPM_CONFIG_LOGLEVEL ; \
NPM_CONFIG_FUND="false" ; export NPM_CONFIG_FUND ; \
NPM_CONFIG_UPDATE_NOTIFIER="false" ; export NPM_CONFIG_UPDATE_NOTIFIER ; \
NO_UPDATE_NOTIFIER="1" ; export NO_UPDATE_NOTIFIER ; \
NPM_CONFIG_COLOR="always" ; export NPM_CONFIG_COLOR ; \

> NPM_CONFIG_INIT_MODULE="%{rpmbuilddir}/.npm-init.js" ; export NPM_CONFIG_INIT_MODULE ; \

Reviewer: ```suggestion
Audit with NPM if package uses bundled Node dependencies. For use in the `%check` section.
```

Reviewer: Why is it not suggesting these for the right lines skull

Reviewer: This is for line 357, adds a period.

#### %npm_audit
Audit with NPM if package uses bundled Node dependencies. For use in the `%check` section

#### %npm_audit_fix
Fix vulnerabilities found using `%npm_audit` Should be used in `%prep` section.

Installed packages with fixed dependencies SHOULD be tested first to make sure the applied fix did not cause any issues.

Available flags:
- `-d` will run with `--dry-run` so you can check what the fix would do.
- `-f` will run with `--force`, ignoring package locks or other requirements set by the module to fix the dependencies. MUST be tested locally and verified to not cause breakage.
- `-o` will omit a dependency or group (such as `dev`) from being fixed.
- `-O` will only fix a specified depency or group.
- `-p` will limit the run to the package lock.

LIMITATION: Cannot be used on modules installed using the `-g` flag (which is done in `%npm_prep`) which use a package lock.

### Bun Macros
[TODO]

#### %__bun_home
Unfortunately, Bun doesn't actually have a way to enforce this yet. However the proposed environment variable is `$BUN_HOME`, so we have this for the future.

> 

Reviewer: ```suggestion
- `-a` can be used to trust all dependencies, otherwise simply list the dependencies you want to trust after the command.
```

Reviewer: Aaaaa these comments
Anyway you spelled dependancies wrong lol

NOTE: Bun audit is a currently a wrapper for NPM audit but lacks the `fix` option, in order to fix vulnerabilities you need to use `%npm_audit_fix`.

You DON'T need to fix vulnerabilities for Node modules that are __not bundled in the final package.__ This means if a vulnerability is found *only* in build dependencies, you are fine to leave them be.

#### %bun_pm_trust
Add Node modules to trusted dependencies so their postinstall scripts may be run. Should only be used if necessary.

Available flags:
- `-a` can be used to trust all dependencies, otherwise simply list the depdencies you want to trust after the command.

Expands to: `%{__bun} pm trust`

### PNPM Macros

#### %_pnpm_home
Where PNPM files and cache are stored.

Expands to: `%{rpmbuilddir}/.pnpm`

#### %_pnpm_store
Where Node modules installed by PNPM usually go. This will usually not be produced when using PNPM as a build tool, but exists for rare builds it may be created in.

Expands to: `%{_pnpm_home}/store`

#### %pnpm_common_envvars
Sets the NPM evironment variables and the PNPM home dir.

Expands to: `%{npm_common_envvars} PNPM_HOME=%{_pnpm_home}`

> #### %__pnpm

Reviewer: ```suggestion
You SHOULD NOT need to fix vulnerabilities for Node modules that are __not bundled in the final package.__ This means if a vulnerability is found *only* in build dependencies, you are fine to leave them be.
```

Reviewer: This is for like 469, replaces DONT with the 'proper' SHOULD NOT
